toolchain/sem_ir/facet_type_info.cpp

```diff
@@ -232,4 +235,50 @@ IdentifiedFacetType::IdentifiedFacetType(
   SortAndDeduplicate(required_interfaces_, RequiredLess);
 }
 
+auto AddCanonicalWitnessesBlock(File& sem_ir,
+                                llvm::SmallVector<InstId>& witnesses)
+    -> InstBlockId {
+  // Small blocks don't need to be sorted.
+  if (witnesses.size() <= 1) {
+    return sem_ir.inst_blocks().AddCanonical(witnesses);
+  }
+
+  llvm::SmallVector<std::pair<SpecificInterface, InstId>> sortable;
+  sortable.reserve(witnesses.size());
+
+  // Produce the sorted order based on the witness's SpecificInterface.
+  for (auto witness_id : witnesses) {
+    auto inst = sem_ir.insts().Get(witness_id);
+    CARBON_KIND_SWITCH(inst) {
+      case CARBON_KIND(ImplWitness witness): {
+        auto table =
+            sem_ir.insts().GetAs<ImplWitnessTable>(witness.witness_table_id);
+        sortable.push_back(
+            {sem_ir.impls().Get(table.impl_id).interface, witness_id});
+        break;
+      }
+      case CARBON_KIND(LookupImplWitness witness): {
+        sortable.push_back({sem_ir.specific_interfaces().Get(
+                                witness.query_specific_interface_id),
+                            witness_id});
+        break;
+      }
+      default:
+        CARBON_FATAL("Unhandled inst: {0}", inst);
+    }
+  }
+  llvm::sort(sortable, [](auto& lhs, auto& rhs) {
```

Comment:
I expect that this is true, but figured I would check -- there can't be cases with the _same_ specific interface IDs, and so this is a total order and not a partial order?